File: app/src/host/figmaActions.ts
/**
 * Running the figma-CLI without a terminal.
 *
 * The panel exists to keep the terminal for Claude, so everything the CLI does on the user's
 * behalf — connecting, the daemon, binding a file, scaffolding the agent rules — runs here as a
 * plain child process: arguments as an array, no shell, output captured and reported in the UI
 * instead of scrolling past Claude's conversation.
 */

import { execFile } from 'child_process';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { resolveCliInvocation, type CliInvocation } from '../lib/cli-command';
import { pathWithShim, shimDir, shimPath, shimScript } from '../lib/cli-shim';
import { loginShellPath } from './ptyManager';

const HOME = os.homedir();
const STATE_DIR = path.join(HOME, '.figma-ds-cli');

/** Written by `figma-cli render`; the undo button reads it and nothing else. */
export const LAST_RENDER_FILE = path.join(STATE_DIR, 'last-render.json');

/** fig-start stores the checkout here, so a user who ran it once needs no panel setting. */
const FIG_START_CONFIG = path.join(HOME, '.figma-cli', 'config.json');

export interface RunResult {
  ok: boolean;
  stdout: string;
  stderr: string;
  /** The single line worth showing: the CLI's last meaningful output, or the failure. */
  message: string;
}

function repoPathFromFigStart(): string {
  try {
    const raw: unknown = JSON.parse(fs.readFileSync(FIG_START_CONFIG, 'utf8'));
    const value = (raw as { repoPath?: unknown }).repoPath;
    return typeof value === 'string' ? value : '';
  } catch {
    return '';
  }
}

/**
 * Where the CLI is, in the order that respects the user's own setup: an explicit setting, then an
 * installed binary, then a checkout — the machine this was built on has only the last one.
 */
export function resolveCli(appRoot: string, configured?: string): CliInvocation {
  return resolveCliInvocation({
    pathDirs: (loginShellPath() ?? process.env.PATH ?? '').split(':').filter(Boolean),
    exists: fs.existsSync,
    configured,
    checkoutDirs: [repoPathFromFigStart(), path.dirname(appRoot)]
  });
}

/** The PATH a spawned CLI needs: node lives in the login shell's PATH, not in launchd's. */
function cliEnv(extra?: Record<string, string>): NodeJS.ProcessEnv {
  const env = { ...process.env, ...extra };
  const loginPath = loginShellPath();
  if (loginPath) env.PATH = loginPath;
  // Inherited from whatever started the app; it would turn a spawned Electron into plain node.
  delete env.ELECTRON_RUN_AS_NODE;
  return env;
}

/** ANSI colours out, empty lines out: the CLI writes for a terminal, this reads in a popover. */
const ANSI = new RegExp(String.fromCharCode(27) + '\\[[0-9;]*[A-Za-z]', 'g');

function lastLine(text: string): string {
  const lines = text
    .replace(ANSI, '')
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean);
  return lines.length > 0 ? lines[lines.length - 1] : '';
}

export function runCli(
  cli: CliInvocation,
  args: string[],
  options: { cwd?: string; env?: Record<string, string>; timeoutMs?: number } = {}
): Promise<RunResult> {
  if (!cli.file) {
    return Promise.resolve({
      ok: false,
      stdout: '',
      stderr: '',
      message: 'figma-cli not found — set "figmaCli" in ~/.figma-ds-cli/panel.json'
    });
  }

  return new Promise((resolve) => {
    execFile(
      cli.file,
      [...cli.args, ...args],
      {
        cwd: options.cwd,
        env: cliEnv(options.env),
        timeout: options.timeoutMs ?? 20_000,
        maxBuffer: 4 * 1024 * 1024
      },
      (error, stdout, stderr) => {
        const ok = !error;
        resolve({
          ok,
          stdout,
          stderr,
          message: ok
            ? lastLine(stdout)
            : lastLine(stderr) || lastLine(stdout) || (error?.message ?? 'failed')
        });
      }
    );
  });
}

export interface OpenFile {
  title: string;
  id: string;
}

/** The open design files, as `figma-cli files` reports them, with Figma's title suffix removed. */
export async function listOpenFiles(cli: CliInvocation): Promise<OpenFile[]> {
  const result = await runCli(cli, ['files'], { timeoutMs: 10_000 });
  if (!result.ok) return [];
  try {
    const parsed: unknown = JSON.parse(result.stdout);
    if (!Array.isArray(parsed)) return [];
    return parsed
      .filter((file): file is { title?: string; id?: string } => Boolean(file))
      .map((file) => ({
        title: String(file.title ?? '')
          .replace(/\s*[–—-]\s*Figma\s*$/u, '')
          .trim(),
        id: String(file.id ?? '')
      }))
      .filter((file) => file.title !== '');
  } catch {
    return [];
  }
}

/**
 * Writes the `figma-cli` launcher the panel's terminals see on their PATH.
 *
 * Only when the CLI runs through node — a genuinely installed binary is already on PATH, and a
 * shim would shadow the real thing. Returns the directory to prepend, or null when there is
 * nothing to do.
 */
export function ensureShim(cli: CliInvocation): string | null {
  if (!cli.entry) return null;

  const dir = shimDir(HOME);
  const file = shimPath(HOME);
  const wanted = shimScript(cli.entry);
  try {
    if (fs.existsSync(file) && fs.readFileSync(file, 'utf8') === wanted) return dir;
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(file, wanted, { mode: 0o755 });
    fs.chmodSync(file, 0o755);
    return dir;
  } catch (error) {
    console.error('[panel] could not write the figma-cli shim:', error);
    return null;
  }
}

export { pathWithShim };

/** A Figma process exists. macOS only — the app ships for macOS, the rest gets an honest false. */
export function isFigmaRunning(): Promise<boolean> {
  if (process.platform !== 'darwin') return Promise.resolve(false);
  return new Promise((resolve) => {
    execFile('/usr/bin/pgrep', ['-x', 'Figma'], (error, stdout) => {
      resolve(!error && stdout.trim() !== '');
    });
  });
}

/** Quits Figma. Only ever called after the user confirmed it in a dialog. */
export function quitFigma(): Promise<void> {
  return new Promise((resolve) => {
    execFile('/usr/bin/pkill', ['-x', 'Figma'], () => resolve());
  });
}

/** The debug port answers — the one probe that says whether Yolo Mode can work at all. */
export async function isCdpReachable(port: number): Promise<boolean> {
  try {
    const response = await fetch(`http://127.0.0.1:${String(port)}/json/version`, {
      signal: AbortSignal.timeout(1500)
    });
    return response.ok;
  } catch {
    return false;
  }
}

/** What the last render created, as the state file holds it. */
export function readLastRender(): string {
  try {
    return fs.readFileSync(LAST_RENDER_FILE, 'utf8');
  } catch {
    return '';
  }
}

export function clearLastRender(): void {
  try {
    fs.unlinkSync(LAST_RENDER_FILE);
  } catch {
    // Already gone is the state we wanted.
  }
}

/** Whether `init-agent` has run in this directory. */
export function hasAgentRules(cwd: string): boolean {
  if (!cwd) return false;
  try {
    return fs.readFileSync(path.join(cwd, 'AGENTS.md'), 'utf8').includes('# Using figma-cli');
  } catch {
    return false;
  }
}

File: swift-host/Sources/FigmaClaudeCore/FigmaActions.swift
import Foundation

/// Running the CLI without a terminal.
///
/// Port of the parts of `app/src/host/figmaActions.ts` the toolbar needs. The panel exists to
/// keep the terminal for Claude, so everything it does on the user's behalf — connecting, the
/// daemon, undoing the last render — runs as a plain child process with its arguments as an
/// array, no shell, output captured and reported in the UI instead of scrolling past Claude's
/// conversation.

public struct CliResult {
    public var ok: Bool
    public var output: String

    public init(ok: Bool, output: String) {
        self.ok = ok
        self.output = output
    }
}

/// Where `fig-start` records which checkout it drove, so a GUI launch can find the CLI even when
/// nothing is installed globally.
private func repoPathFromConfig() -> String? {
    let file = NSHomeDirectory() + "/.figma-cli/config.json"
    guard let data = FileManager.default.contents(atPath: file),
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          let path = object["repoPath"] as? String, !path.isEmpty else { return nil }
    return path
}

/// The CLI as this machine can actually run it: configured, on PATH, or a checkout in place.
public func resolveCli(appRoot: String, configured: String? = nil) -> CliInvocation {
    let path = LoginShellPath.resolve() ?? ProcessInfo.processInfo.environment["PATH"] ?? ""
    let dirs = path.split(separator: ":").map(String.init)

    // FigmaClaude ships inside the CLI repo, so the app's own parent is a candidate.
    var checkouts: [String] = []
    if let repo = repoPathFromConfig() { checkouts.append(repo) }
    checkouts.append((appRoot as NSString).deletingLastPathComponent)
    checkouts.append(NSHomeDirectory() + "/figma-cli")

    return resolveCliInvocation(pathDirs: dirs, configured: configured, checkoutDirs: checkouts)
}

/// Runs the CLI once and captures what it said. Arguments as an array, never a shell string:
/// a Figma file name with a quote in it must not become a shell injection.
@discardableResult
public func runCli(_ cli: CliInvocation, _ arguments: [String],
                   cwd: String? = nil, timeout: TimeInterval = 60) -> CliResult {
    guard cli.isUsable else {
        return CliResult(ok: false, output: "figma-cli not found — install it or set figmaCli in panel.json")
    }

    let path = LoginShellPath.resolve() ?? ProcessInfo.processInfo.environment["PATH"] ?? ""
    let resolved = cli.file.contains("/") ? cli.file : (whichOnPath(cli.file, path: path) ?? cli.file)

    let process = Process()
    process.executableURL = URL(fileURLWithPath: resolved)
    process.arguments = cli.args + arguments
    if let cwd { process.currentDirectoryURL = URL(fileURLWithPath: cwd) }
    var environment = ProcessInfo.processInfo.environment
    environment["PATH"] = path
    process.environment = environment

    let pipe = Pipe()
    process.standardOutput = pipe
    process.standardError = pipe

    do {
        try process.run()
    } catch {
        return CliResult(ok: false, output: "could not start \(resolved): \(error.localizedDescription)")
    }

    let data = pipe.fileHandleForReading.readDataToEndOfFile()
    process.waitUntilExit()

    // ANSI out: this text goes into a dialog, not a terminal.
    let output = String(decoding: data, as: UTF8.self)
        .replacingOccurrences(of: "\u{001B}\\[[0-9;]*[A-Za-z]", with: "", options: .regularExpression)
        .trimmingCharacters(in: .whitespacesAndNewlines)

    return CliResult(ok: process.terminationStatus == 0, output: output)
}

/// Undoes the last render over the daemon's `/exec` route, so the button works on a machine
/// where the CLI is only a checkout. The ids come from `last-render.json` alone — nothing is
/// searched for or guessed, so it can never touch anything else on the canvas.
public func undoLastRender() -> String {
    let raw = try? String(contentsOfFile: lastRenderFile, encoding: .utf8)
    let nodes = parseLastRender(raw)
    guard !nodes.isEmpty else { return undoMessage(removed: 0, names: []) }

    guard let reply = daemonEvaluate(buildUndoEval(ids: nodes.map(\.id)), timeout: 20),
          let data = reply.data(using: .utf8),
          let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else {
        // The daemon answers with the object itself rather than a JSON string for some routes;
        // reporting "nothing" here would claim the canvas is clean when it may not be.
        return "Could not reach the daemon — is Figma connected?"
    }

    let removed = object["removed"] as? Int ?? 0
    let names = object["names"] as? [String] ?? []
    if removed > 0 { try? FileManager.default.removeItem(atPath: lastRenderFile) }
    return undoMessage(removed: removed, names: names)
}
